[Writing Tools] Reverting a proofreading session may not restore the original text

pxlcoder opened the pull request:
#### 8dfd5f755284bb2823e064842a90bfef185d406e
<pre>
[Writing Tools] Reverting a proofreading session may not restore the original text
<a href="https://bugs.webkit.org/show_bug.cgi?id=280955">https://bugs.webkit.org/show_bug.cgi?id=280955</a>
<a href="https://rdar.apple.com/137233726">rdar://137233726</a>

Reviewed by Richard Robinson.

Reverting a proofreading session is performed by iterating document markers in
reverse order and restoring the original text specified in the marker data.

However, when adding a marker for a suggestion with multiple words, more than
one marker may get added, as `DocumentMarkerController` adds one marker per
text range. These text ranges are determined using
`DocumentMarkerController::collectTextRanges`.

For most marker types, markers for adjacent text ranges are merged. However,
since `shouldInsertAsSeparateMarker` returns true for proofreading markers,
the extra markers are not merged. Finally, this results in an issue when
performing a revert, since additional text will be restored, due to the presence
of additional markers.

Fix by allowing proofreading markers to be merged if their IDs are the same.

* Source/WebCore/dom/DocumentMarkerController.cpp:
(WebCore::shouldInsertAsSeparateMarker):
(WebCore::canMergeMarkers):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
(TEST(WritingTools, ProofreadingRevertWithMultiwordSuggestions)):

Canonical link: <a href="https://commits.webkit.org/284775@main">https://commits.webkit.org/284775@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/53c96fd050ea9be845a9f87957f1d73048cdfd82

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70491 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49897 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23256 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/74580 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/21669 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/72608 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57697 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21509 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55856 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/14326 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73557 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45382 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60768 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36318 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42041 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/18204 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/20030 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/63973 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18548 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76299 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14717 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/17786 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/63578 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14760 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60835 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63512 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11556 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/5194 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10794 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45699 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/466 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/46773 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/48050 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46515 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
```diff
@@ -620,6 +620,64 @@ static void checkColor(WebCore::CocoaColor *color, std::optional<ColorExpectatio
+
+        EXPECT_WK_STREQ(originalText, contexts.firstObject.attributedText.string);
+
+        RetainPtr suggestions = [NSMutableArray array];
```
I think this would look a bit nicer as an array literal, but doesn't really matter

pxlcoder replied:
Changed.

rr-codes commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
```diff
@@ -620,6 +620,64 @@ static void checkColor(WebCore::CocoaColor *color, std::optional<ColorExpectatio
+
+        [webView waitForNextPresentationUpdate];
+
+        NSString *proofreadContents = [NSString stringWithFormat:@"%@\n%@", proofreadText, signature];
```
should we use a RetainPtr here 😅 (genuine question)

pxlcoder replied:
Sure.

pxlcoder committed:
[Writing Tools] Reverting a proofreading session may not restore the original text
https://bugs.webkit.org/show_bug.cgi?id=280955
rdar://137233726

Reviewed by Richard Robinson.

Reverting a proofreading session is performed by iterating document markers in
reverse order and restoring the original text specified in the marker data.

However, when adding a marker for a suggestion with multiple words, more than
one marker may get added, as `DocumentMarkerController` adds one marker per
text range. These text ranges are determined using
`DocumentMarkerController::collectTextRanges`.

For most marker types, markers for adjacent text ranges are merged. However,
since `shouldInsertAsSeparateMarker` returns true for proofreading markers,
the extra markers are not merged. Finally, this results in an issue when
performing a revert, since additional text will be restored, due to the presence
of additional markers.

Fix by allowing proofreading markers to be merged if their IDs are the same.

* Source/WebCore/dom/DocumentMarkerController.cpp:
(WebCore::shouldInsertAsSeparateMarker):
(WebCore::canMergeMarkers):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
(TEST(WritingTools, ProofreadingRevertWithMultiwordSuggestions)):

Canonical link: https://commits.webkit.org/284775@main
